csharp/src/Microsoft.ML.OnnxRuntime/DisposableNamedOnnxValue.cs

```diff
@@ -0,0 +1,177 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Numerics.Tensors;
+
+
+namespace Microsoft.ML.OnnxRuntime
+{
+    public interface IDisposableReadOnlyCollection<T>: IReadOnlyCollection<T>, IDisposable
+    {
+
+    }
+
+    internal class DisposableList<T> : List<T>, IDisposableReadOnlyCollection<T>
+        where T: IDisposable
+    {
+
+        #region IDisposable Support
+        private bool disposedValue = false; // To detect redundant calls
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!disposedValue)
+            {
+                if (disposing)
+                {
+                    // TODO: dispose managed state (managed objects).
+                    for (int i = 0; i < this.Count; i++)
+                    {
+                        this[i].Dispose();
+                    }
+                    this.Clear();
+                }
+
+                // TODO: free unmanaged resources (unmanaged objects) and override a finalizer below.
+                // TODO: set large fields to null.
+
+                disposedValue = true;
+            }
+        }
+
+        ~DisposableList()
+        {
+            // Do not change this code. Put cleanup code in Dispose(bool disposing) above.
+            Dispose(false);
```

Comment:
>Dispose(false); [](start = 12, length = 15)

This should have an assert or logging statement in case when we got here and the object was not properly disposed. This is the only proper use of the finalizers.